File: backend/routes/assignment.js
const express = require("express");
const Posts = require("../models/assignment");
const multer = require("multer");
const mongoose = require("mongoose");

const path = require("path");

const router = express.Router();

// Multer storage configuration for assignment upload
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    if (file.fieldname === "assignment") {
      cb(null, path.join(__dirname, "../Assignments")); // Correct path for storing assignments
    } else {
      cb(new Error("Invalid fieldname"), null);
    }
  },
  filename: (req, file, cb) => {
    cb(null, Date.now() + path.extname(file.originalname)); // Unique file name with timestamp
  },
});

const upload = multer({ storage: storage });


// Get all assignments with pagination
router.get("/assignment", async (req, res) => {
  try {
    const page = parseInt(req.query.page) || 1;
    const limit = parseInt(req.query.limit) || 8;
    const startIndex = (page - 1) * limit;

    const posts = await Posts.find()
      .sort({ createdAt: -1 }) // Sort by createdAt in descending order
      .skip(startIndex)
      .limit(limit)
      .exec(); // Get the latest posts first
    const totalPosts = await Posts.countDocuments();

    return res.status(200).json({
      success: true,
      existingPosts: posts,
      totalPosts,
    });
  } catch (error) {
    return res.status(400).json({ error: error.message });
  }
});

// Get a specific assignment by ID
router.get("/assignment/:id", async (req, res) => {
  try {
    const postId = req.params.id;
    const post = await Posts.findById(postId).exec();
    if (!post) {
      return res
        .status(404)
        .json({ success: false, message: "Assignment not found" });
    }
    return res.status(200).json({ success: true, post });
  } catch (error) {
    return res.status(400).json({ error: error.message });
  }
});



router.get("/assignment/get/:assignment_id", async (req, res) => {
  try {
    const { assignment_id } = req.params; // Get assignment_id from the URL

    // Find the video by video_id
    const Assignment = await Posts.findOne({ assignment_id: assignment_id });

    if (!Assignment) {
      return res.status(404).json({ error: "assignment not found" });
    }

    return res.status(200).json(Assignment); // Return the video details
  } catch (error) {
    return res.status(400).json({ error: error.message });
  }
});

router.put(
  "/assignment/update/:assignment_id",
  upload.single("assignment"), // Handles the assignment file
  async (req, res) => {
    try {
      const { assignment_id } = req.params;
      const { card_id, assignment_name } = req.body;

      // Validate input
      if (!assignment_id || !card_id || !assignment_name) {
        return res.status(401).json({ error: "All fields are required" });
      }

      // Find the existing assignment by assignment_id
      const existingAssignment = await Posts.findOne({ assignment_id });
      if (!existingAssignment) {
        return res.status(404).json({ error: "Assignment not found" });
      }

      let updatedAssignmentPath = existingAssignment.assignment; // Default to the current file path

      // If a new file is uploaded, handle the file update
      if (req.file) {
        const newFilePath = req.file.filename;

        // Delete the old file
        const oldFilePath = path.join(
          __dirname,
          "../Assignments",
          existingAssignment.assignment
        );
        fs.unlink(oldFilePath, (err) => {
          if (err) {
            console.error("Error deleting old file:", err);
          } else {
            console.log("Old file deleted successfully:", oldFilePath);
          }
        });

        // Update the file path
        updatedAssignmentPath = newFilePath;
      }

      // Update the assignment in the database
      existingAssignment.card_id = card_id;
      existingAssignment.assignment_name = assignment_name;
      existingAssignment.assignment = updatedAssignmentPath;

      // Save the updated record
      await existingAssignment.save();

      return res
        .status(200)
        .json({ success: "Assignment updated successfully" });
    } catch (error) {
      console.error("Error updating assignment:", error);
      return res.status(500).json({ error: error.message });
    }
  }
);

// Delete assignment route
router.delete("/assignment/delete/:assignment_id", async (req, res) => {
  try {
    const { assignment_id } = req.params;

    // Validate input
    if (!assignment_id) {
      return res.status(400).json({ error: "assignment_id is required" });
    }

    // Find the assignment by assignment_id
    const assignment = await Posts.findOne({ assignment_id });
    if (!assignment) {
      return res.status(404).json({ error: "Assignment not found" });
    }

    // Delete the file from storage if it exists
    const filePath = path.join(
      __dirname,
      "../Assignments",
      assignment.assignment
    );
    fs.unlink(filePath, (err) => {
      if (err) {
        console.error("Error deleting file:", err);
      } else {
        console.log("File deleted successfully:", filePath);
      }
    });

    // Remove the database record
    await Posts.deleteOne({ assignment_id });
    return res.status(200).json({ success: "Assignment deleted successfully" });
  } catch (error) {
    console.error("Error deleting assignment:", error);
    return res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: backend/routes/assignment.py
import os
from fastapi import APIRouter, UploadFile, File, Form, Depends, HTTPException
from sqlalchemy.orm import Session
from datetime import date
from services.s3_service import delete_file, upload_file
from database import get_db
from models.assignment import Assignment
from schemas.assignment import AssignmentResponse
from typing import List



router = APIRouter(prefix="/assignments", tags=["Assignments"])


# CREATE ASSIGNMENT (Teacher)
@router.post("/", response_model=AssignmentResponse)
async def create_assignment(
    module_id: int = Form(...),
    title: str = Form(...),
    full_marks: int = Form(...),   # ✅ ADD THIS
    description: str = Form(None),
    file: UploadFile = File(None),
    open_date: date = Form(None),
    close_date: date = Form(None),
    allow_download: bool = Form(True),
    db: Session = Depends(get_db),
):
    file_path = None

    # Stream the file directly to S3 if it exists
    if file:
        try:
            file_path = upload_file(file, folder="assignments")
        except Exception as e:
            raise HTTPException(
                status_code=500, 
                detail=f"Failed to upload file to cloud storage: {str(e)}"
            )

    assignment = Assignment(
        module_id=module_id,
        title=title,
        full_marks=full_marks,   # ✅ SAVE IT
        description=description,
        file_path=file_path,     # Saves the S3 Object HTTPS URL string
        open_date=open_date,
        close_date=close_date,
        allow_download=allow_download,
    )

    db.add(assignment)
    db.commit()
    db.refresh(assignment)

    return assignment


# GET ASSIGNMENTS BY MODULE ID
@router.get("/module/{module_id}", response_model=List[AssignmentResponse])
def get_assignments_by_module(
    module_id: int,
    db: Session = Depends(get_db),
):
    assignments = (
        db.query(Assignment)
        .filter(Assignment.module_id == module_id)
        .all()
    )

    if not assignments:
        raise HTTPException(
            status_code=404,
            detail="No assignments found for this module"
        )

    return assignments


# UPDATE ASSIGNMENT
@router.put("/update/{assignment_id}", response_model=AssignmentResponse)
async def update_assignment(
    assignment_id: int,
    module_id: int = Form(None),
    title: str = Form(None),
    full_marks: int = Form(None),
    description: str = Form(None),
    file: UploadFile = File(None),
    open_date: date = Form(None),
    close_date: date = Form(None),
    allow_download: bool = Form(None),
    db: Session = Depends(get_db),
):
    assignment = (
        db.query(Assignment)
        .filter(Assignment.id == assignment_id)
        .first()
    )

    if not assignment:
        raise HTTPException(
            status_code=404,
            detail="Assignment not found"
        )

    # Handle S3 file updates cleanly
    if file:
        # Delete old file from S3 if it has an active URL
        if assignment.file_path:
            delete_file(assignment.file_path)

        # Upload new replacement file to S3
        try:
            assignment.file_path = upload_file(file, folder="assignments")
        except Exception as e:
            raise HTTPException(
                status_code=500, 
                detail=f"Failed to upload updated file to cloud storage: {str(e)}"
            )

    # Update other model fields
    if module_id is not None:
        assignment.module_id = module_id

    if title is not None:
        assignment.title = title

    if full_marks is not None:
        assignment.full_marks = full_marks

    if description is not None:
        assignment.description = description

    if open_date is not None:
        assignment.open_date = open_date

    if close_date is not None:
        assignment.close_date = close_date

    if allow_download is not None:
        assignment.allow_download = allow_download

    db.commit()
    db.refresh(assignment)

    return assignment


# DELETE ASSIGNMENT
@router.delete("/{assignment_id}")
def delete_assignment(
    assignment_id: int,
    db: Session = Depends(get_db),
):
    assignment = db.query(Assignment).filter(Assignment.id == assignment_id).first()

    if not assignment:
        raise HTTPException(status_code=404, detail="Assignment not found")

    # If the assignment contains an S3 URL asset, delete it from the bucket
    if assignment.file_path:
        delete_file(assignment.file_path)

    db.delete(assignment)
    db.commit()

    return {"message": "Assignment deleted successfully"}


# GET ASSIGNMENT BY ID
@router.get("/{assignment_id}", response_model=AssignmentResponse)
def get_assignment_by_id(
    assignment_id: int,
    db: Session = Depends(get_db),
):
    assignment = (
        db.query(Assignment)
        .filter(Assignment.id == assignment_id)
        .first()
    )

    if not assignment:
        raise HTTPException(
            status_code=404,
            detail="Assignment not found"
        )

    return assignment